File: mordisco-api/src/main/java/utn/back/mordiscoapi/enums/EstadoPedido.java
package utn.back.mordiscoapi.enums;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

import java.util.EnumSet;
import java.util.Set;

@Getter
@RequiredArgsConstructor
public enum EstadoPedido {
    // Estados comunes
    PENDIENTE("Pendiente", "Pedido creado, esperando confirmación del restaurante"),
    EN_PREPARACION("En Preparación", "El restaurante está preparando tu pedido"),

    // Estados para RETIRO_POR_LOCAL
    LISTO_PARA_RETIRAR("Listo para Retirar", "Tu pedido está listo, puedes pasar a retirarlo"),

    // Estados para DELIVERY
    LISTO_PARA_ENTREGAR("Listo para Entregar", "Pedido listo para que el repartidor lo retire"),
    EN_CAMINO("En Camino", "El repartidor está llevando tu pedido"),

    // Estado final común
    COMPLETADO("Completado", "Pedido finalizado exitosamente"),

    // Estado de cancelación
    CANCELADO("Cancelado", "Pedido cancelado");

    private final String displayName;
    private final String descripcion;

    // Estados válidos por tipo de entrega
    private static final Set<EstadoPedido> ESTADOS_RETIRO = EnumSet.of(
            PENDIENTE, EN_PREPARACION, LISTO_PARA_RETIRAR, COMPLETADO, CANCELADO
    );

    private static final Set<EstadoPedido> ESTADOS_DELIVERY = EnumSet.of(
            PENDIENTE, EN_PREPARACION, LISTO_PARA_ENTREGAR, EN_CAMINO, COMPLETADO, CANCELADO
    );

    /**
     * Verifica si este estado es válido para un tipo de entrega
     */
    public boolean esValidoParaTipoEntrega(TipoEntrega tipoEntrega) {
        return switch (tipoEntrega) {
            case RETIRO_POR_LOCAL -> ESTADOS_RETIRO.contains(this);
            case DELIVERY -> ESTADOS_DELIVERY.contains(this);
        };
    }

    /**
     * Obtiene los estados válidos para un tipo de entrega
     */
    public static Set<EstadoPedido> obtenerEstadosValidosPara(TipoEntrega tipoEntrega) {
        return switch (tipoEntrega) {
            case RETIRO_POR_LOCAL -> ESTADOS_RETIRO;
            case DELIVERY -> ESTADOS_DELIVERY;
        };
    }

    /**
     * Verifica si se puede cambiar de este estado a otro, considerando el tipo de entrega
     */
    public boolean puedeCambiarA(EstadoPedido nuevoEstado, TipoEntrega tipoEntrega) {
        // No se puede cambiar si ya está completado o cancelado
        if (this == COMPLETADO || this == CANCELADO) {
            return false;
        }

        // Validar que ambos estados sean válidos para el tipo de entrega
        if (!this.esValidoParaTipoEntrega(tipoEntrega) || !nuevoEstado.esValidoParaTipoEntrega(tipoEntrega)) {
            return false;
        }

        // Definir transiciones válidas según tipo de entrega
        return switch (tipoEntrega) {
            case RETIRO_POR_LOCAL -> puedeCambiarRetiroLocal(nuevoEstado);
            case DELIVERY -> puedeCambiarDelivery(nuevoEstado);
        };
    }

    private boolean puedeCambiarRetiroLocal(EstadoPedido nuevoEstado) {
        return switch (this) {
            case PENDIENTE -> nuevoEstado == EN_PREPARACION || nuevoEstado == CANCELADO;
            case EN_PREPARACION -> nuevoEstado == LISTO_PARA_RETIRAR || nuevoEstado == CANCELADO;
            case LISTO_PARA_RETIRAR -> nuevoEstado == COMPLETADO || nuevoEstado == CANCELADO;
            default -> false;
        };
    }

    private boolean puedeCambiarDelivery(EstadoPedido nuevoEstado) {
        return switch (this) {
            case PENDIENTE -> nuevoEstado == EN_PREPARACION || nuevoEstado == CANCELADO;
            case EN_PREPARACION -> nuevoEstado == LISTO_PARA_ENTREGAR || nuevoEstado == CANCELADO;
            case LISTO_PARA_ENTREGAR -> nuevoEstado == EN_CAMINO || nuevoEstado == CANCELADO;
            case EN_CAMINO -> nuevoEstado == COMPLETADO || nuevoEstado == CANCELADO;
            default -> false;
        };
    }

    /**
     * Verifica si el pedido puede ser cancelado en este estado
     */
    public boolean esCancelable() {
        return this == PENDIENTE || this == EN_PREPARACION || this == LISTO_PARA_RETIRAR || this == LISTO_PARA_ENTREGAR;
    }

    /**
     * Obtiene el siguiente estado lógico según el tipo de entrega
     */
    public EstadoPedido obtenerSiguienteEstado(TipoEntrega tipoEntrega) {
        return switch (tipoEntrega) {
            case RETIRO_POR_LOCAL -> switch (this) {
                case PENDIENTE -> EN_PREPARACION;
                case EN_PREPARACION -> LISTO_PARA_RETIRAR;
                case LISTO_PARA_RETIRAR -> COMPLETADO;
                default -> this;
            };
            case DELIVERY -> switch (this) {
                case PENDIENTE -> EN_PREPARACION;
                case EN_PREPARACION -> LISTO_PARA_ENTREGAR;
                case LISTO_PARA_ENTREGAR -> EN_CAMINO;
                case EN_CAMINO -> COMPLETADO;
                default -> this;
            };
        };
    }

    /**
     * Verifica si este es un estado final
     */
    public boolean esFinal() {
        return this == COMPLETADO || this == CANCELADO;
    }

    /**
     * Verifica si este estado requiere acción del repartidor
     */
    public boolean requiereRepartidor() {
        return this == LISTO_PARA_ENTREGAR || this == EN_CAMINO;
    }
}

File: mordisco-front/src/app/shared/models/enums/estado-pedido.ts
import { TipoEntrega } from "./tipo-entrega";

export enum EstadoPedido {
  PENDIENTE = 'PENDIENTE',
  EN_PREPARACION = 'EN_PREPARACION',
  LISTO_PARA_RETIRAR = 'LISTO_PARA_RETIRAR',
  LISTO_PARA_ENTREGAR = 'LISTO_PARA_ENTREGAR',
  EN_CAMINO = 'EN_CAMINO',
  COMPLETADO = 'COMPLETADO',
  CANCELADO = 'CANCELADO'
}

export const ESTADO_PEDIDO_LABELS: Record<EstadoPedido, string> = {
  [EstadoPedido.PENDIENTE]: 'Pendiente',
  [EstadoPedido.EN_PREPARACION]: 'En Preparación',
  [EstadoPedido.LISTO_PARA_RETIRAR]: 'Listo para Retirar',
  [EstadoPedido.LISTO_PARA_ENTREGAR]: 'Listo para Entregar',
  [EstadoPedido.EN_CAMINO]: 'En Camino',
  [EstadoPedido.COMPLETADO]: 'Completado',
  [EstadoPedido.CANCELADO]: 'Cancelado'
};

export const ESTADO_PEDIDO_COLORS: Record<EstadoPedido, string> = {
  [EstadoPedido.PENDIENTE]: 'bg-yellow-100 text-yellow-700',
  [EstadoPedido.EN_PREPARACION]: 'bg-blue-100 text-blue-700',
  [EstadoPedido.LISTO_PARA_RETIRAR]: 'bg-green-100 text-green-700',
  [EstadoPedido.LISTO_PARA_ENTREGAR]: 'bg-purple-100 text-purple-700',
  [EstadoPedido.EN_CAMINO]: 'bg-indigo-100 text-indigo-700',
  [EstadoPedido.COMPLETADO]: 'bg-emerald-100 text-emerald-700',
  [EstadoPedido.CANCELADO]: 'bg-red-100 text-red-700'
};

export const ESTADO_PEDIDO_ICONS: Record<EstadoPedido, string> = {
  [EstadoPedido.PENDIENTE]: 'schedule',
  [EstadoPedido.EN_PREPARACION]: 'restaurant',
  [EstadoPedido.LISTO_PARA_RETIRAR]: 'shopping_bag',
  [EstadoPedido.LISTO_PARA_ENTREGAR]: 'inventory_2',
  [EstadoPedido.EN_CAMINO]: 'local_shipping',
  [EstadoPedido.COMPLETADO]: 'check_circle',
  [EstadoPedido.CANCELADO]: 'cancel'
};

/**
 * Obtiene los estados válidos según el tipo de entrega
 */
export function getEstadosValidosPorTipo(tipoEntrega: TipoEntrega): EstadoPedido[] {
  if (tipoEntrega === TipoEntrega.RETIRO_POR_LOCAL) {
    return [
      EstadoPedido.PENDIENTE,
      EstadoPedido.EN_PREPARACION,
      EstadoPedido.LISTO_PARA_RETIRAR,
      EstadoPedido.COMPLETADO,
      EstadoPedido.CANCELADO
    ];
  } else {
    return [
      EstadoPedido.PENDIENTE,
      EstadoPedido.EN_PREPARACION,
      EstadoPedido.LISTO_PARA_ENTREGAR,
      EstadoPedido.EN_CAMINO,
      EstadoPedido.COMPLETADO,
      EstadoPedido.CANCELADO
    ];
  }
}

/**
 * Obtiene el siguiente estado lógico según el tipo de entrega
 */
export function getSiguienteEstado(
  estadoActual: EstadoPedido, 
  tipoEntrega: TipoEntrega
): EstadoPedido | null {
  if (tipoEntrega === TipoEntrega.RETIRO_POR_LOCAL) {
    switch (estadoActual) {
      case EstadoPedido.PENDIENTE: return EstadoPedido.EN_PREPARACION;
      case EstadoPedido.EN_PREPARACION: return EstadoPedido.LISTO_PARA_RETIRAR;
      case EstadoPedido.LISTO_PARA_RETIRAR: return EstadoPedido.COMPLETADO;
      default: return null;
    }
  } else {
    switch (estadoActual) {
      case EstadoPedido.PENDIENTE: return EstadoPedido.EN_PREPARACION;
      case EstadoPedido.EN_PREPARACION: return EstadoPedido.LISTO_PARA_ENTREGAR;
      case EstadoPedido.LISTO_PARA_ENTREGAR: return EstadoPedido.EN_CAMINO;
      case EstadoPedido.EN_CAMINO: return EstadoPedido.COMPLETADO;
      default: return null;
    }
  }
}

/**
 * Verifica si se puede cambiar de un estado a otro
 */
export function puedeCambiarEstado(
  estadoActual: EstadoPedido,
  nuevoEstado: EstadoPedido,
  tipoEntrega: TipoEntrega
): boolean {
  // No puede cambiar si ya está completado o cancelado
  if (estadoActual === EstadoPedido.COMPLETADO || 
      estadoActual === EstadoPedido.CANCELADO) {
    return false;
  }

  // Puede cancelar en cualquier momento (excepto si ya está completado o cancelado, verificado arriba)
  if (nuevoEstado === EstadoPedido.CANCELADO) {
    return true;
  }

  // Solo puede avanzar al siguiente estado lógico
  const siguienteEstado = getSiguienteEstado(estadoActual, tipoEntrega);
  return nuevoEstado === siguienteEstado;
}

/**
 * Obtiene la descripción del estado según el contexto
 */
export function getEstadoDescripcion(
  estado: EstadoPedido,
  tipoEntrega: TipoEntrega,
  rol: 'CLIENTE' | 'RESTAURANTE' | 'REPARTIDOR'
): string {
  const descripciones: Record<string, Record<string, string>> = {
    [EstadoPedido.PENDIENTE]: {
      CLIENTE: 'Esperando confirmación del restaurante',
      RESTAURANTE: 'Nuevo pedido - requiere tu confirmación',
      REPARTIDOR: 'Pedido pendiente de preparación'
    },
    [EstadoPedido.EN_PREPARACION]: {
      CLIENTE: 'Tu pedido se está preparando',
      RESTAURANTE: 'Preparando pedido',
      REPARTIDOR: 'El restaurante está preparando el pedido'
    },
    [EstadoPedido.LISTO_PARA_RETIRAR]: {
      CLIENTE: '¡Tu pedido está listo! Puedes pasar a retirarlo',
      RESTAURANTE: 'Pedido listo - esperando que el cliente lo retire',
      REPARTIDOR: 'N/A'
    },
    [EstadoPedido.LISTO_PARA_ENTREGAR]: {
      CLIENTE: 'Tu pedido está listo para ser entregado',
      RESTAURANTE: 'Pedido listo - esperando repartidor',
      REPARTIDOR: 'Pedido listo para retirar del restaurante'
    },
    [EstadoPedido.EN_CAMINO]: {
      CLIENTE: '¡Tu pedido está en camino!',
      RESTAURANTE: 'El repartidor está llevando el pedido',
      REPARTIDOR: 'Entregando pedido al cliente'
    },
    [EstadoPedido.COMPLETADO]: {
      CLIENTE: tipoEntrega === TipoEntrega.RETIRO_POR_LOCAL 
        ? 'Pedido retirado exitosamente' 
        : 'Pedido entregado exitosamente',
      RESTAURANTE: 'Pedido completado',
      REPARTIDOR: 'Pedido entregado exitosamente'
    },
    [EstadoPedido.CANCELADO]: {
      CLIENTE: 'Pedido cancelado',
      RESTAURANTE: 'Pedido cancelado',
      REPARTIDOR: 'Pedido cancelado'
    }
  };

  return descripciones[estado]?.[rol] || 'Estado desconocido';
}